base/sparse/sparsematrix.jl

```diff
@@ -1084,18 +1084,67 @@ end # macro
 (.-)(A::Number, B::SparseMatrixCSC) = A .- full(B)
 ( -)(A::Array , B::SparseMatrixCSC) = A  - full(B)
 
+# multiplication and division by scalars need to be careful about 0, Inf, NaN
+# corner cases where we might need to return dense data (as a SparseMatrixCSC
+# for type stability)
+function densify_with_default(A::SparseMatrixCSC, spvals, defaultvalue)
+    # return a SparseMatrixCSC C with the same dimensions as A, structural
+    # nonzero values spvals in the same locations that A has structural
+    # nonzeros, and nonzero value defaultvalue everywhere else
+    m, n = size(A)
+    Arowval = A.rowval
+    Acolptr = A.colptr
+    Cnnz = m * n
+    Cnzval = fill(defaultvalue, Cnnz)
+    Crowval = similar(Arowval, Cnnz)
+    Ccolptr = similar(Acolptr)
+    Ccolptr[1] = 1
+    for col = 1:n
+        Ccolptr[col+1] = 1 + col * m
+        Crowval[Ccolptr[col] : Ccolptr[col+1]-1] = 1:m
+        for k in nzrange(A, col)
+            Cnzval[sub2ind((m, n), Arowval[k], col)] = spvals[k]
+        end
+    end
+    return SparseMatrixCSC(m, n, Ccolptr, Crowval, Cnzval)
+end
+
 (.*)(A::AbstractArray, B::AbstractArray) = broadcast_zpreserving(MulFun(), A, B)
-(.*)(A::SparseMatrixCSC, B::Number) = SparseMatrixCSC(A.m, A.n, copy(A.colptr), copy(A.rowval), A.nzval .* B)
-(.*)(A::Number, B::SparseMatrixCSC) = SparseMatrixCSC(B.m, B.n, copy(B.colptr), copy(B.rowval), A .* B.nzval)
+function (.*)(A::SparseMatrixCSC, B::Number)
+    if isfinite(B)
+        SparseMatrixCSC(A.m, A.n, copy(A.colptr), copy(A.rowval), A.nzval .* B)
+    else
+        densify_with_default(A, A.nzval .* B, zero(eltype(A)) .* B)
```

Comment:
At least theoretically, now that we have fast anonymous functions you could pass `B` and `(x,y) ->x .* y`, and thus skip the intermediate storage for `A.nzval .* B`.